### ISSUE TITLE
[Feature Request] more flexible splitting of Tensors

### ISSUE THREAD

**Issue by clavichord93** [pytorch/pytorch#3223]
Currently, `torch.split` only splits a Tensor into equal-size (if possible) chunks. But sometimes Tensors need to be split into more flexible chunks.

For example, splitting a Tensor of size (256, 256, 32, 32) into two chunks of sizes (256, 32, 32, 32) and (256, 224, 32, 32) along dim 1, like `tf.split` in TensorFlow.

So can this be added into PyTorch standard operators?

**Comment by ptrblck** [comment]
Hi,

I tried to implement your desired functionality, since I think it might be quite useful.

Here is the code snippet:

    def size_splits(tensor, split_sizes, dim=0):
        """Splits the tensor according to chunks of split_sizes.
        
        Arguments:
            tensor (Tensor): tensor to split.
            split_sizes (list(int)): sizes of chunks
            dim (int): dimension along which to split the tensor.
        """
        if dim < 0:
            dim += tensor.dim()
        
        dim_size = tensor.size(dim)
        if dim_size != torch.sum(torch.Tensor(split_sizes)):
            raise KeyError("Sum of split sizes exceeds tensor dim")
        
        splits = torch.cumsum(torch.Tensor([0] + split_sizes), dim=0)[:-1]
    
        return tuple(tensor.narrow(int(dim), int(start), int(length)) 
            for start, length in zip(splits, split_sizes))

Here are some tests:

    a = torch.randn(20, 10, 2, 2)
    
    split_sizes = [2, 2, 6]
    dim = 1
    res = size_splits(tensor=a, split_sizes=split_sizes, dim=dim)
    
    for r, split_size in zip(res, split_sizes):
        assert r.size(dim) == split_size
    
    try:
        res = size_splits(tensor=a, split_sizes=split_sizes, dim=0)
    except KeyError as e: 
        print(e)
    
    split_sizes = [5, 5, 10]
    dim = 0
    res = size_splits(tensor=a, split_sizes=split_sizes, dim=dim)
    
    for r, split_size in zip(res, split_sizes):
        assert r.size(dim) == split_size

I would like to create a pull request, if others find this function useful as well.

Greets!

**Comment by clavichord93** [comment]
This is great! Efficiency is crucial for this operation, so how efficient is your implementation?

**Comment by ptrblck** [comment]
You can add this code to test it on your machine:

    print(timeit.timeit('size_splits(tensor=torch.randn(200, 10, 2, 2), split_sizes=[5]*40, dim=0)',
                  setup='from __main__ import size_splits\nimport torch', number=10000))
    print(timeit.timeit('torch.split(tensor=torch.randn(200, 10, 2, 2), split_size=5, dim=0)',
                  setup='import torch', number=10000))

This yields for 
size_splits: 6.991s
torch.split: 6.704s

I'm comparing equally sized chunks, since that is what `torch.split` is doing.

**Comment by clavichord93** [comment]
Wow, it looks pretty good. I will use it in my code. Thank you!

**Comment by samuela** [comment]
Thanks @ptrblck ! This is extremely useful.

Nit: `raise KeyError("Sum of split sizes exceeds tensor dim")` should be `raise KeyError("Sum of split sizes does not equal tensor dim")`.

**Comment by ptrblck** [comment]
Thanks for the suggestion @samuela !

I am not sure how to proceed? Should I just create a pull request or does it need any further discussion from the core devs? 
For example, I'm not sure if this function should be implemented in the `split` method (with a condition) or as a standalone method.
What do you think?

**Comment by clavichord93** [comment]
@ptrblck @samuela Maybe there should be a ‘ValueError’ instead of a ‘KeyError’?

**Comment by samuela** [comment]
@clavichord93 true true

**Comment by anands-repo** [comment]
@ptrblck Thanks. This is a great feature.

Quick question: Is the following 
` return tuple(tensor.narrow(int(dim), int(start), int(length)) 
        for start, length in zip(splits, split_sizes))`

faster than running it in a for loop

For example, when run on a GPU, would these statements be executed in parallel just because the multiple "torch.narrow"s are in a single statement?

**Comment by ptrblck** [comment]
@anands-repo `split` was moved a while ago to ATen, so that the Python implementation is not used anymore.
[ATen implementation](https://github.com/pytorch/pytorch/blob/821f8bfc2fe9b57712a40723ae4983712f568fc7/aten/src/ATen/native/TensorShape.cpp#L712).

**Comment by mmiakashs** [comment]
@ptrblck Thanks for this feature.
Is it possible to extend this feature to split a tensor on a dimension with overlapping split sequence, somewhat like torch.nn.Unfold . However, torch.nn.Unfold is only limited to split batched images, like 4d tensor. But is there any fastest and easiest way in PyTorch to split batch of sequence images (B x L x C x W x H) with overlapping splits on a particular dimension say L? I think it is very necessary feature in now days for multi-dimensional sequence data. Thanks in advance.

**Comment by ptrblck** [comment]
@mmiakashs With `tensor.unfold` you can directly specify which dimension you would like to unfold:
```python
x = torch.randn(1, 2, 3, 4, requires_grad=True)
patches = x.unfold(dimension=3, size=2, step=1)
patches.mean().backward()
print(x.grad)
```
Would that work for you?

**Comment by mmiakashs** [comment]
@ptrblck Thanks a lot for your quick response.
Yes, that's what I am looking for. But according to the docs unfold is only working for 4d tensor only. I was working with 5d tensor (B x L x C x W x H). Will it work, if I merge the batch(B) and sequence length(L) in one dimension then apply the unfold. Finally, unmerge the batch(B) and sequence length(L)

[Docs: torch.nn.unfold](https://pytorch.org/docs/stable/nn.html#unfold)
> Currently, only 4-D input tensors (batched image-like tensors) are supported.

**Comment by ptrblck** [comment]
You approach might work. Alternatively, try `tensor.unfold` instead of `torch.nn.Unfold`.

**Comment by mmiakashs** [comment]
> You approach might work. Alternatively, try `tensor.unfold` instead of `torch.nn.Unfold`.

Thanks a lot, it works

**Comment by bubas3000** [comment]
@ptrblck this feature is amazing. Is it possible to get the max over split?
For example:
a = [0 , 1,2,3,4]
torch.split(a,[1,2,2]) = ([0],[1,2],[3,4])
I would like the output of max_split to be [0,2,3]
Thanks in advance

**Comment by bubas3000** [comment]
> @ptrblck this feature is amazing. Is it possible to get the max over split?
> For example:
> a = [0 , 1,2,3,4]
> torch.split(a,[1,2,2]) = ([0],[1,2],[3,4])
> I would like the output of max_split to be [0,2,3]
> Thanks in advance

Oops, the output should be [0,2,4] of course...
